src/scenic/core/shapes.py

```diff
@@ -0,0 +1,151 @@
+
+from abc import ABC, abstractmethod
+
+import trimesh
+from trimesh.transformations import translation_matrix, quaternion_matrix, concatenate_matrices
+import numpy
+
+from scenic.core.vectors import Orientation
+from scenic.core.utils import cached_property, loadMesh
+
+###################################################################################################
+# Abstract Classes and Utilities
+###################################################################################################
+
+class Shape(ABC):
+    """An abstract base class for Scenic shapes.
+
+    Represents a physical shape in Scenic. Does not encode position or orientation,
+    which are handled by the `Region` class. Does contain dimension information, which
+    is used as a default value by any `Object` with this shape and can be overwritten.
+
+    Args:
+        dimensions: The raw (before scaling) dimensions of the shape. If dimensions
+          and scale are both specified the dimensions are first set by dimensions, and then
+          scaled by scale.
+        scale: Scales all the dimensions of the shape by a multiplicative factor.
+          If dimensions and scale are both specified the dimensions are first set by dimensions,
+          and then scaled by scale.
+    """
+    def __init__(self, dimensions, scale):
+        # Store values
+        self.raw_dimensions = dimensions
+        self.scale = scale
+        self.dimensions = tuple(dim * self.scale for dim in self.raw_dimensions)
+        self.width = self.dimensions[0]
+        self.length = self.dimensions[1]
+        self.height = self.dimensions[2]
+
+    @cached_property
+    def containsCenter(self):
+        """Whether or not this object contains its central point"""
+        pq = trimesh.proximity.ProximityQuery(self.mesh)
+        region_distance = pq.signed_distance([(0,0,0)])[0]
+
+        return region_distance > 0
+
+    @property
+    @abstractmethod
+    def mesh(self):
+        pass
+
+    @property
+    @abstractmethod
+    def isConvex(self):
+        pass
+
+###################################################################################################
+# 3D Shape Classes
+###################################################################################################
+
+class MeshShape(Shape):
+    """ A Shape subclass defined by a Trimesh object.
+
+    Args:
+        mesh: A trimesh.Trimesh mesh object.
+        dimensions: The raw (before scaling) dimensions of the shape. If dimensions
+          and scale are both specified the dimensions are first set by dimensions, and then
+          scaled by scale.
+        scale: Scales all the dimensions of the shape by a multiplicative factor.
+          If dimensions and scale are both specified the dimensions are first set by dimensions,
+          and then scaled by scale.
+        initial_rotation: A 3-tuple containing the yaw, pitch, and roll respectively to apply when loading
```

Comment:
Since this parameter is available to the user, we should use `toOrientation` here and so accept not only triples but headings and `Orientation` objects.